### PR TITLE
Add support for multiple unit systems and for custom formatters/parsers to the QuantityFormatter

### PR DESCRIPTION
**Pull request by bsteinbk**
Add API calls to allow one or more FormatterParserSpecsProvider(s) to be registered with QuantityFormatter. Each FormatterParserSpecsProvider will register custom FormatSpecs and ParserSpecs for a custom QuantityType specified by a unique string id.  This will allow custom formatting and parsing logic  for quantity values. This feature was adapted from work done by the Civil team.

Add support for 4 unit systems that match the four provide by Presentation Manager: Metric, Imperial, US Survey, and US Customary. There are now default quantity formats defined for each unit system. Only one unit system can be "active" the API call ImodelApp.quantityFormatter.setActiveUnitSystem is used to activate the unit system cache formatter and parser specs and  emit onActiveUnitSystemChanged event.

Example usage of new capabilities added to ui-test-app and ui-test-extension.

### CODE REVIEW

**Review by DanEastBentley — APPROVED** [review]

**Review by NancyMcCallB — APPROVED** [review]

File: core/frontend-devtools/src/tools/ChangeUnitsTool.ts [hunk 1]
```diff
@@ -19,12 +21,21 @@ export class ChangeUnitsTool extends Tool {
   public static get minArgs() { return 0; }
   public static get maxArgs() { return 1; }
 
-  public run(useMetric?: boolean): boolean {
+  // support boolean for backwards compatibility - string allows options to four unit systems
+  public run(useMetric?: boolean | string): boolean {
```
Comment (pmconne) [inline core/frontend-devtools/src/tools/ChangeUnitsTool.ts:25]: If you are changing what arguments a key-in accepts, please update the README accordingly.

**Review by pmconne — CHANGES_REQUESTED** [review]

File: core/frontend-devtools/src/tools/ChangeUnitsTool.ts (hunk 1 above)
Comment (bsteinbk) [reply core/frontend-devtools/src/tools/ChangeUnitsTool.ts]: I reverted my change to support passing in a string as the key-in "fdt metric" would look quit odd if you keyed in "fdt metric imperial" or "fdt metric usSurvey" to set it to a different unit system.

File: core/frontend/src/QuantityFormatter.ts [hunk 2]
```diff
@@ -83,382 +66,579 @@ const unitData: UnitDefinition[] = [
...
+  protected _activeFormatSpecsByType = new Map<QuantityTypeKey, FormatterSpec>();
+  protected _activeParserSpecsByType = new Map<QuantityTypeKey, ParserSpec>();
+  protected _overrideFormatPropsByQuantityType = new Map<QuantityTypeKey, OverrideFormatEntry>();
+  protected _formatSpecProviders = new Array<FormatterParserSpecsProvider>();
+
+  /** Called after the active unit system is changed.
+   * The useImperial argument should not be relied on now that multiple systems are supported. The system parameter
+   * will contain the system name being activated.
+   */
+  public readonly onActiveUnitSystemChanged = new BeUiEvent<{ useImperial?: boolean, system?: string }>();
```
Comment (a-gagnon) [inline core/frontend/src/QuantityFormatter.ts:525]: Is there a reason why the arguments of the event are optional? The `useImperial` use to be required in onActiveUnitSystemChanged. (Same comment for onQuantityFormatsChanged.)

Comment (bsteinbk) [reply core/frontend/src/QuantityFormatter.ts:525]: As the comment states, the user should now use the 'system' argument instead. I kept the useImperial argument so I would not break any existing users, but wouldn't removing "?" from system be a breaking change?.

Comment (pmconne) [reply core/frontend/src/QuantityFormatter.ts:525]: Might be cleaner to deprecate old event and add new event with desired args.

Comment (bsteinbk) [reply core/frontend/src/QuantityFormatter.ts:525]: Good idea! That is what I will do.

**Review by pmconne — APPROVED** [review]

**Review by calebmshafer — CHANGES_REQUESTED** [review]
Looks good overall.  A couple minor comments

File: docs/changehistory/NextVersion.md [hunk 3]
```diff
@@ -191,6 +191,8 @@ New
 
 - Changed `highlightProps?: HighlightedRecordProps` property to `highlight?: HighlightingComponentProps` on [PropertyRendererProps]($ui-components) interface. To react to this change, simply rename `highlightProps` -> `highlight`.
 
+- The methods fromJson and toJson from [Format]($quantity) have been renamed to fromJSON and toJSON respectively.
```
[the remark this answers is not in the document]
Comment (bsteinbk) [reply docs/changehistory/NextVersion.md]: Updated to remove link notation and add info about changing parseIntoQuantityValue to parseToQuantityValue

File: core/quantity/src/ParserSpec.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+   *  @param unitsProvider The units provider is used to look up unit definitions and provide conversion information for converting between units.
+   *  @param outUnit The unit the value to be formatted. This unit is often referred to as persistence unit.
+   */
+  public static async create(format: Format, unitsProvider: UnitsProvider, outUnit: UnitProps): Promise<ParserSpec> {
+    const conversions = await Parser.createUnitConversionSpecsForUnit(unitsProvider, outUnit);
+    return new ParserSpec(outUnit, format, conversions);
+  }
+
+  /** Do the parsing. Done this way to allow Custom Parser Specs to parse custom formatted strings into their quantities. */
+  public parseIntoQuantityValue(inString: string): ParseResult {
```
[the remark this answers is not in the document]
Comment (bsteinbk) [reply core/quantity/src/ParserSpec.ts]: Updated method names in both ParserSpec and QuantityFormatter alpha classes.

**Review by calebmshafer — APPROVED** [review]

**Pull request merged**